crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -563,33 +557,43 @@ where
             // Handle shielded input
             // The following boundary conditions must be satisfied
             // 1. Zero transparent input
-            // 2. The spend descriptions' anchors are valid
-            // 3. The convert descriptions's anchors are valid
-            // 4. The nullifiers provided by the transaction have not been
-            // revealed previously (even in the same tx) and no unneeded
-            // nullifier is being revealed by the tx
-            if let Some(transp_bundle) = shielded_tx.transparent_bundle() {
-                if !transp_bundle.vin.is_empty() {
-                    tracing::debug!(
-                        "Transparent input to a transaction from the masp \
-                         must be 0 but is {}",
-                        transp_bundle.vin.len()
-                    );
-                    return Ok(false);
-                }
+            // 2. At least one shielded input
+            // 3. The spend descriptions' anchors are valid
+            // 4. The convert descriptions's anchors are valid
+            if shielded_tx
+                .transparent_bundle()
+                .is_some_and(|bundle| !bundle.vin.is_empty())
+            {
+                tracing::debug!(
+                    "No transparent inputs are allowed when the source is masp",
+                );
+                return Ok(false);
+            }
+
+            if !shielded_tx
+                .sapling_bundle()
+                .is_some_and(|bundle| !bundle.shielded_spends.is_empty())
+            {
+                return Ok(false);
             }
 
             if !(self.valid_spend_descriptions_anchor(&shielded_tx)?
-                && self.valid_convert_descriptions_anchor(&shielded_tx)?
-                && self.valid_nullifiers_reveal(keys_changed, &shielded_tx)?)
+                && self.valid_convert_descriptions_anchor(&shielded_tx)?)
             {
                 return Ok(false);
             }
         }
 
         // The transaction must correctly update the note commitment tree
-        // in storage with the new output descriptions
-        if !self.valid_note_commitment_update(&shielded_tx)? {
+        // in storage with the new output descriptions and also reveal the
+        // nullifiers correctly (only if needed) NOTE: these two checks
+        // validate the keys that the transaction write in storage and therefore
+        // must be done regardless of the type of transaction (shielding,
+        // shielded, unshielding) since a malicious tx could try to write keys
+        // in an invalid way
+        if !(self.valid_note_commitment_update(&shielded_tx)?
+            || self.valid_nullifiers_reveal(keys_changed, &shielded_tx)?)
```

Comment:
Shouldn't this OR be an AND? We want both the note commitment update and the nullifier reveal to be valid. No?

Reply:
... and we should definitely have tests for this if we don't already